File: Data Structures & Algorithms/eating-bananas/submission-8.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int maxRate = piles[0], lowestRate = 1;
        for (int i = 0; i < piles.length; i++) {
            maxRate = Math.max(maxRate, piles[i]);
        }

        int res = maxRate;
        while (lowestRate <= maxRate) {
            int rate = lowestRate + ((maxRate - lowestRate) / 2);
            
            int time = 0;    
            for (int i = 0; i < piles.length; i++) {
                time += (int) Math.ceil((double) piles[i] / rate);
                if (time > h) {
                    break;
                }
            }

            if (time <= h) {
                res = rate;
                maxRate = rate - 1; // try to find smaller rate
            } else {
                lowestRate = rate + 1;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-8.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0

        l = 0
        for r in range(len(prices)):
            profit = prices[r] - prices[l]

            if profit > 0:
                res = max(res, profit)
            elif profit < 0:
                l = r
        
        return res